yellow-paper/docs/contract-deployment/classes.md

```diff
@@ -16,114 +16,158 @@ Read the following discussions for additional context:
-- Computes the `contract_class_id` as the hash of the `ContractClass` object. <!-- TODO: describe the exact hashing computation. -->
-- Emits the computed `contract_class_id` as a nullifier.
-  - Note: The [private kernel circuit](../circuits/private-kernel-initial.md) will 'silo' <!-- TODO: link to relevant section --> this so-called "inner nullifier" (the `contract_class_id` itself) with the contract address of the `ContractClassRegisterer` contract to produce a so-called "outer nullifier" (the `contract_class_id_nullifier`). So nodes will need to reconstruct the `contract_class_id` from the `NewContractClassRegistration` event, and then attempt to reconstruct the "outer nullifier", to validate both the `contract_class_id` and that the canonical `ContractClassRegisterer` was indeed the source of the event.
+- Assert that `packed_bytecode` is valid according to the definition in the [Public VM section](../public-vm/bytecode-validation-circuit.md#packed-bytecode-representation).
+- Computes the `contract_class_id` as [defined above](#class-identifier).
+- Emits the resulting `contract_class_id` as a nullifier to prevent the same class from being registered again.
+- Emits an unencrypted event `ContractClassRegistered` with the contents of the contract class.
 
-<!-- Q: should we write some pseudocode, or do you think these bullet points suffice? -->
+In pseudocode:
 
-Upon seeing a `NewContractClassRegistration` event in a mined transaction, nodes are expected to store the contract class, so they can retrieve it when executing a public function for that class.
+```
+function register(
+  artifact_hash: Field,
+  private_functions: { selector: u32, vk_hash: Field }[],
+  packed_bytecode: Field[],
+) 
+  assert is_valid_packed_bytecode(packed_bytecode)
 
-Note that emitting the `contract_class_id` as a nullifier (the `contract_class_id_nullifier`), instead of as an entry in the note hashes tree, allows public functions to prove non-existence of a class, which is required to support public contract instance deployments.
+  version = 1
+  private_function_leaves = private_functions.map(fn => pedersen([fn.function_selector as Field, fn.vk_hash], GENERATOR__FUNCTION_LEAF))
+  private_functions_root = merkleize(private_functions)
+  bytecode_commitment = calculate_commitment(packed_bytecode)
+  contract_class_id = pedersen([version, artifact_hash, private_functions_root, bytecode_commitment], GENERATOR__CLASS_IDENTIFIER)
 
-<!-- [Mike review, for @spalladino][cspell:disable-line]
-TODO: Perhaps we should write the details of this Registration contract in a similar way to ERC interface specifications? E.g. specify all state variables, functions, event names of the Registration contract. (Or are there no other functions, other than `register`?).
--->
+  emit_nullifier contract_class_id
+  emit_event ContractClassRegistered(contract_class_id, version, artifact_hash, private_functions, packed_bytecode)
```

Comment:
Changed the `register` function so it just broadcasts the `private_functions_root` instead (since we just need to make public the public bytecode), and added separate functions to individually broadcast bytecode for specific private or unconstrained functions. Also added a recommendation for devs to broadcast `compute_hash_and_nullifier` at the time they register a class (which is easy to do all in the same tx since we have multicall built in!).

> It is strongly recommended for developers registering new classes to broadcast the code for `compute_hash_and_nullifier`, so any private message recipients have the code available to process their incoming notes. However, the `ContractClassRegisterer` contract does not enforce this during registration, since it is difficult to check the multiple signatures for `compute_hash_and_nullifier` as they may evolve over time to account for new note sizes.